### PR TITLE
feat(db): implement `BusinessProfileInterface` for `MockDb`

### LINKED ISSUE #2039: Implement `BusinessProfileInterface` for `MockDb`
Spin off of #172 

Implement the missing implementation of MockDb for business profiles.
https://github.com/juspay/hyperswitch/blob/b8501f6ae6c0af0804d7f1e7cf9f50ebd303bf94/crates/router/src/db/business_profile.rs#L114

### PR DESCRIPTION
**Pull request by dalprahcd**
Implement the missing implementation of MockDb for business profiles.

Closes #2039

## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/db/business_profile.rs [hunk 1]
```diff
@@ -114,38 +114,118 @@ impl BusinessProfileInterface for Store {
 impl BusinessProfileInterface for MockDb {
     async fn insert_business_profile(
         &self,
-        _business_profile: business_profile::BusinessProfileNew,
+        business_profile: business_profile::BusinessProfileNew,
     ) -> CustomResult<business_profile::BusinessProfile, errors::StorageError> {
-        Err(errors::StorageError::MockDbError)?
+        let mut business_profiles = self.business_profiles.lock().await;
+
+        let business_profile_insert = business_profile::BusinessProfile {
```
Comment (dalprahcd) [inline crates/router/src/db/business_profile.rs:121]: Quick question, could this be changed to the following?

```rust
let business_profile_insert = business_profile::BusinessProfile {
    ..business_profile
};
```

Comment (SanchithHegde) [reply crates/router/src/db/business_profile.rs:121]: I'd expect it to not work, using the struct update syntax needs both the source and destination structs to be the same type.

**Review by SanchithHegde — APPROVED** [review]
Looks good to me!

Thanks for the PR, @dalprahcd!

File: crates/router/src/db/business_profile.rs [hunk 2]
```diff
@@ -114,38 +114,118 @@ impl BusinessProfileInterface for Store {
...
+        {
+            Some(index) => {
+                _ = business_profiles.remove(index);
+                Ok(true)
+            }
+            None => Err(errors::StorageError::ValueNotFound(
+                "cannot find business profile to delete".to_string(),
+            )
+            .into()),
+        }
```
Comment (ThisIsMani) [inline crates/router/src/db/business_profile.rs:213]: Same here, you can improve this by changing this into like this:

```suggestion
        let index = business_profiles
            .iter()
            .position(|bp| bp.profile_id == profile_id && bp.merchant_id == merchant_id)
            .ok_or(errors::StorageError::ValueNotFound(
                    "cannot find business profile to delete".to_string(),
                )
                .into()
            )?;
            business_profiles.remove(index);
            Ok(true)
```

File: crates/router/src/db/business_profile.rs [hunk 3]
```diff
@@ -114,38 +114,118 @@ impl BusinessProfileInterface for Store {
...
+        match business_profiles
+            .iter()
+            .position(|bp| bp.profile_id == profile_id && bp.merchant_id == merchant_id)
+        {
+            Some(index) => {
+                _ = business_profiles.remove(index);
+                Ok(true)
+            }
+            None => Err(errors::StorageError::ValueNotFound(
+                "cannot find business profile to delete".to_string(),
```
Comment (ThisIsMani) [inline crates/router/src/db/business_profile.rs:210]: This error message can also be improved.

File: crates/router/src/db/business_profile.rs (hunk 1 above)
Comment (ThisIsMani) [reply crates/router/src/db/business_profile.rs:121]: Generally conversions between types is achieved by `From` and `Into` traits. Refer [this](https://doc.rust-lang.org/rust-by-example/conversion/from_into.html).

**Review by ThisIsMani — COMMENTED** [review]
Few suggestions

File: crates/router/src/db/business_profile.rs (hunk 1 above)
Comment (dalprahcd) [reply crates/router/src/db/business_profile.rs]: Changed to use `impl From<BusinessProfileNew> for BusinessProfile`

File: crates/router/src/db/business_profile.rs [hunk 4]
```diff
@@ -114,38 +114,118 @@ impl BusinessProfileInterface for Store {
...
+            .business_profiles
+            .lock()
+            .await
+            .iter()
+            .find(|business_profile| business_profile.profile_id == profile_id)
+        {
+            Some(business_profile) => Ok(business_profile.clone()),
+            None => {
+                return Err(errors::StorageError::ValueNotFound(
+                    "business profile not found".to_string(),
```
[the remark this answers is not in the document]
Comment (dalprahcd) [reply crates/router/src/db/business_profile.rs]: Changed to `"No business profile found for profile_id = {profile_id}"`

File: crates/router/src/db/business_profile.rs [hunk 5]
```diff
@@ -114,38 +114,118 @@ impl BusinessProfileInterface for Store {
...
+            .map(|bp| {
+                let business_profile_updated =
+                    business_profile_update.apply_changeset(current_state);
+                *bp = business_profile_updated.clone();
+                business_profile_updated
+            }) {
+            Some(business_profile) => Ok(business_profile),
+            None => {
+                return Err(errors::StorageError::ValueNotFound(
+                    "cannot find business profile to update".to_string(),
```
[the remark this answers is not in the document]
Comment (dalprahcd) [reply crates/router/src/db/business_profile.rs]: Changed to "No business profile found for profile_id = {profile_id}"

**Comment by dalprahcd** [comment]
I have addressed the suggestions.

File: crates/router/src/db/business_profile.rs [hunk 6]
```diff
@@ -114,38 +114,89 @@ impl BusinessProfileInterface for Store {
...
+            .find(|bp| bp.profile_id == current_state.profile_id)
+            .map(|bp| {
+                let business_profile_updated =
+                    business_profile_update.apply_changeset(current_state);
+                *bp = business_profile_updated.clone();
+                business_profile_updated
+            })
+            .ok_or(
+                errors::StorageError::ValueNotFound(
+                    "No business profile found for profile_id = {profile_id}".to_string(),
```
Comment (ThisIsMani) [inline crates/router/src/db/business_profile.rs:163]: This seems incorrect, can you please check it.

Comment (dalprahcd) [reply crates/router/src/db/business_profile.rs:163]: Sorry, I am still learning Rust, the code builds and clippy does not throw warnings. Could you explain more what seems incorrect?

Comment (ThisIsMani) [reply crates/router/src/db/business_profile.rs:163]: No worries, I meant error message that you have here is incorrect as `profile_id` that you have used in error message is neither a local variable nor passed in args.

```suggestion
                    "No business profile found for profile_id = {current_state.profile_id}".to_string(),
```

File: crates/router/src/db/business_profile.rs [hunk 7]
```diff
@@ -114,38 +114,89 @@ impl BusinessProfileInterface for Store {
...
     ) -> CustomResult<business_profile::BusinessProfile, errors::StorageError> {
-        Err(errors::StorageError::MockDbError)?
+        self.business_profiles
+            .lock()
+            .await
+            .iter()
+            .find(|business_profile| business_profile.profile_id == profile_id)
+            .ok_or(
+                errors::StorageError::ValueNotFound(
+                    "No business profile found for profile_id = {profile_id}".to_string(),
```
Comment (ThisIsMani) [inline crates/router/src/db/business_profile.rs:138]: Use `format!` marco here too.
```suggestion
                    format!("No business profile found for profile_id = {}", profile_id),
```

File: crates/router/src/db/business_profile.rs [hunk 8]
```diff
@@ -114,38 +114,89 @@ impl BusinessProfileInterface for Store {
...
+        merchant_id: &str,
     ) -> CustomResult<bool, errors::StorageError> {
-        Err(errors::StorageError::MockDbError)?
+        let mut business_profiles = self.business_profiles.lock().await;
+        let index = business_profiles
+            .iter()
+            .position(|bp| bp.profile_id == profile_id && bp.merchant_id == merchant_id)
+            .ok_or::<errors::StorageError>(
+                errors::StorageError::ValueNotFound(
+                    "No business profile found for profile_id = {profile_id} and merchant_id = {merchant_id}".to_string(),
```
Comment (ThisIsMani) [inline crates/router/src/db/business_profile.rs:180]: Here too. Please use the `format!` macro.

**Comment by dalprahcd** [comment]
Done. I have changed to use `format!()` for error messages.

**Review by ThisIsMani — APPROVED** [review]
Thanks for the fixes!

**Review by SanchithHegde — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]
Thanks for the PR @dalprahcd!

**Pull request merged**